src/packages/core/src/server.ts

```diff
@@ -149,25 +182,36 @@ export default class Server {
     if (this.#status === Status.opening) {
       // if opening
       throw new Error(`Cannot close server while it is opening.`);
-    } else if (this.#status & Status.closed) {
-      // if closed or closing
-      throw new Error(`Server is already closed or closing.`);
+    } else if ((this.#status & Status.closingOrClosed) !== 0) {
+      // if closing or closed
+      throw new Error(`Server is already closing or closed.`);
     }
 
-    const _listenSocket = this.#listenSocket;
     this.#status = Status.closing;
+
+    // clean up the websocket objects
+    const _listenSocket = this.#listenSocket;
     this.#listenSocket = void 0;
     // close the socket to prevent any more connections
-    uWS.us_listen_socket_close(_listenSocket);
+    if (_listenSocket !== null) {
+      uWS.us_listen_socket_close(_listenSocket);
+    }
     // close all the connected websockets:
-    const ws = this.#websocketServer;
-    if (ws) {
-      ws.close();
+    if (this.#websocketServer !== null) {
+      this.#websocketServer.close();
     }
 
     // and do all http cleanup, if any
-    this.#httpServer.close();
-    await this.#connector.close();
+    if (this.#httpServer !== null) {
+      this.#httpServer.close();
+    }
+
+    // cleanup the connector, provider, etc.
+    if (this.#connector !== null) {
+      await this.#connector.close();
+    }
+
+
```

Comment:
Correct me if I'm wrong, but I think it's safe to remove these two guards, as well as the `_listenSocket` guard above, as these conditions should always evaluate to `true` here.

If the code changes in a way that causes these branches to evaluate to `false`, I'd like for it to throw an exception (implicit is fine), and not be guarded away, as when the server is `open`, these values must absolutely be non-null, or things are likely broken in some other subtle way.

Reply:
> as these conditions should always evaluate to `true` here.

I disagree with this. I don't see any evidence in the code that ensures they'll always be true.

For example, you could call `Server.listen()` and `uWS.App()` in `Server.initialize()` could throw, causing you to catch in the error in `listen()` which will call this function; `this.#httpSever = new HttpServer(...)` would never have been called.

If you tried to run this through typescript without the check, it would succeed because you don't have strict null checks turned on. If you had these turned on, typescript would tell you it could not ensure it was properly set. I definitely believe these guard if statements are necessary. Typescript would enforce them or a bang/optional-chain operator had you have strict null checks enabled.

> If the code changes in a way that causes these branches to evaluate to false, I'd like for it to throw an exception (implicit is fine), and not be guarded away, as when the server is open, these values must absolutely be non-null, or things are likely broken in some other subtle way.

This would potentially circumvent the error that is being caught in `listen()` and you'd be throwing a "could not initialize properly" error instead of the actual error that caused initialize to fail. This goes back to your assumption that these values are always non-null at this context, which is not the case. Not to be purposely repetitive, they could be not null in the event initialization fails, in which case would be handled and thrown in the `.catch(...)` logic at line 173.

Reply:
Ah, I missed the new call to `this.close()` in the catch. I think we need to move the `if (callbackIsFunction) callback!(err);` check on line `166` to the new catch block.